text/0079-cdk-2.0.md

```diff
@@ -0,0 +1,337 @@
+
+## Major Version Rollout Tooling
+
+To aid in this phased release, some features will need to be added to JSII and other cdk dev tools. These features will help us with rolling out major versions in the future. Whenever a new breaking release occurs, deprecated APIs will be hidden and feature flags will be reset. When we are going through the next cycle of breaking changes, in this case v3.0, then code related to deprecated APIs and feature flags introduced during the preceding major version cycle, v1.0 in this case, can be removed since they no longer will be supported.
+
+#### [JSII Prerelease Versioning Awareness]()
+
+In order to support releasing the `aws-cdk-lib` with a prerelease version number, JSII needs to parse the version number of the module its building and recognize it as a "prerelease" and convert it to whatever format is equivalent in the target languages package manager.
+
+#### [JSII Submodule Support](https://github.com/aws/jsii/issues/1286)
+
+JSII currently doesn't handle namespacing of exports. Support for this will be needed to allow building the `aws-cdk-lib` module in all of the cdk supported target languages. Though the initial work is complete, some open questions remain about the scope of this feature. Namely, in order to make migrating to the new module structure as easy as possible, [namespace renaming]() could potentially allow C#, Python, and Java users to not have to change the structure of their import statements.
+
+#### [`aws-cdk-lib` Version Unlocking]()
+
+Since `aws-cdk-lib` will have a version number differing from the one used by lerna to version all of the other packages in the repo until the "stable" phase, changes to the align-version and bump scripts will have to be made in the repo to handle this.
+
+#### [Deprecation Aware Builds]()
+
+A strategy for selectively including or excluding apis marked `deprecated` during build time. This can be performed at the JSII level and potentially can be linked to specific version numbers with additional information in the `deprecated` annotation. JSII can warn when APIs marked for deprecation within a specific major version release are being included in the build. For example:
+
+An api annotated like so:
+```typescript
+  /**
+  * @deprecated-v1
+  */
+  function myDeprecatedApi() {}
+```
+
+When inside a module versioned `1.x.x`
+`jsii` - builds normally, no warning
+
+when inside a module versioned `2.x.x` or greater
+`jsii` - build fails with error stating that deprecated APIs are included and points to the corresponding lines of code. It will suggest usage of `--no-deprecations` option to exlcude these APIs.
+`jsii --no-deprecations` - builds without warning, but excludes the above method and others marked deprecated during the v1.0 cycle
```

Comment:
Sorry, I'm not following everything that is said here. Can you clarify better on the text here? It might read better if you start with the use case and then state the flags to use for that use case.

Are you saying that there is now an option to continue to keep deprecated APIs in the new version?

Reply:
No the `--no-deprecations` option is what you use when you want to remove deprecated APIs from the build you are performing. When you build without it, and JSII sees deprecated APIs from the `current - 1` version cycle, it warns you saying that you are in fact including APIs that shouldn't be.